stdlib/public/core/SmallString.swift

```diff
@@ -241,22 +241,44 @@ extension _SmallString {
   fileprivate mutating func withMutableCapacity(
     _ f: (UnsafeMutableRawBufferPointer) throws -> Int
   ) rethrows {
-    let len = try withUnsafeMutableBytes(of: &self._storage) {
-      (rawBufPtr: UnsafeMutableRawBufferPointer) -> Int in
-      let len = try f(rawBufPtr)
-      UnsafeMutableRawBufferPointer(
-        rebasing: rawBufPtr[len...]
-      ).initializeMemory(as: UInt8.self, repeating: 0)
-      return len
-    }
-    if len == 0 {
+    let len = try withUnsafeMutableBytes(of: &_storage, f)
+
+    if len <= 0 {
+      _debugPrecondition(len == 0)
       self = _SmallString()
       return
     }
-    _internalInvariant(len <= _SmallString.capacity)
+    zeroTrailingBytes(of: &_storage, from: len)
+    self = _SmallString(leading: _storage.0, trailing: _storage.1, count: len)
+  }
+}
 
-    let (leading, trailing) = self.zeroTerminatedRawCodeUnits
-    self = _SmallString(leading: leading, trailing: trailing, count: len)
+@inline(__always)
+private func zeroTrailingBytes(
+  of storage: inout _SmallString.RawBitPattern, from len: Int
+) {
+  _internalInvariant(len > 0)
+  _internalInvariant(len <= _SmallString.capacity)
+  if len <= 0 {
+    storage.0 = 0
+    storage.1 = 0
+  } else if len <= 8 {
+    let mask0 = (UInt64(bitPattern: ~0) &>> (8 &* (8 &- len)))
+#if _endian(little)
+    storage.0 &= mask0
+#else //FIXME: Verify this on big-endian architecture
+    #error("Verify correctness on big-endian architecture")
+    storage.0 &= mask0.littleEndian
+#endif
```

Comment:
We could safely move the masking out of the `#if` to simplify this a little.  The two `#if`s could then be merged and moved to the top.

```swift
#if _endian(big)
  #error("Verify correctness on big-endian architecture") // FIXME
#endif
    ...
    storage.0 &= mask0.littleEndian
    ...
    storage.1 &= mask1.littleEndian
```

More importantly, how extensive are the tests for this stuff? (Are the tests enabled on all platforms, including 32-bit ones and Android/AArch64? Do we have tests for the len > 8 && len <= _SmallString.capacity case?)

Reply:
I added a test of the `zeroTrailingBytes` function and removed the conditional compilation statements. I left the `FIXME`s.